pylabrobot/resources/trough.py

```diff
@@ -25,3 +42,29 @@ def __init__(
       category=category,
       model=model
     )
+    self.material_z_thickness = material_z_thickness
+    self.true_dz = true_dz
```

Comment:
what is `true_dz`?

Reply:
The distance between the bottom of a plate/trough and the lower side of the well_material/trough_well_material.

I didn't want to name it `dz` yet because we haven't fixed `dz` in `Plate` yet.

But this information is needed if someone wanted to use this trough anywhere that is not a `TroughCarrierSite` with a pedestal. It is the same situation as with `PlateCarrierSite`.

Reply:
`dz` in plate is just a parameter that will determine the .location.z of wells.

I think since you're building this from scratch now, we might as well do it right from the start, I don't see a reason to sync up with plate. I hope `true` is something we should not have to use, because that implies there is a `false` somewhere. Probably that's true right now, but not something we should bake into the library like that.

Perhaps `through_base_to_container_base` or something could work?

Reply:
> `dz` in plate is just a parameter that will determine the .location.z of wells.

As we discussed in [[PLR Carrier z-Height Issue](https://labautomation.io/t/plr-carrier-z-height-issue/3056)](https://labautomation.io/t/plr-carrier-z-height-issue/3056/35) this is an error, resulting from the incorrect VENUS-to-PLR  transfer of `ThicknessOfContainerBase` to `dz`, which in VENUS is actually the distance between bottom/base of plate_skirt to outside_bottom_of_well_base).
I am confident that VENUS is not using `plate_skirt_base_to_outside_bottom_of_well_base` because it is too long, and instead simply called it `dz` in direct analogy to `dx` and `dy`.

> I hope `true` is something we should not have to use, because that implies there is a `false` somewhere.

That is correct. The current `well.dz` is `false` as we know it is actually `well.thickness_of_container_base` / `well.well_z_thickness`.

But I see your point about immediately starting with the correct attribute names for troughs because we are starting from scratch with their definition and don't have to carefully adapt them like we currently have to do with `Plate` because `Plate` is used by every user so far.

Reply:
So `dz` is actually only a parameter of `create_equally_spaced_2d`. As with `dx` and `dy`, it is not an attribute of ItemizedResource (or Plate). In that method call, the name is self-explanatory.

I think for the attribute here, it should really be named something descriptive like `through_base_to_container_base`.

Reply:
> As with `dx` and `dy`, it [`dz`] is not an attribute of ItemizedResource (or Plate).

Yes, I don't think we should keep it like this. You have seen the computational load this created for `PlateAdapter`, even though in the real world these three attributes _are_ attributes of the physical plate. `Plate` should enable access to all the attributes that define a physical plate.
But I will raise this issue - if I find the time - again in a separate discussion.

Okay, then you are suggesting we handle trough completely differently to well-plate even though it is exactly the same. As long as it is consistent I don't mind.
I have now changed `Trough.true_dz` to `Trough.through_base_to_container_base` and the PR is ready to merge from my side

Reply:
> Okay, then you are suggesting we handle trough completely differently to well-plate even though it is exactly the same.

I don't think Trough and Well are exactly the same tbh. `Well`s have a z-location to the `Plate` (so there are two layers of resources), whereas with a `Trough` there is only one layer. The `dz` parameter is simply a shorthand for creating many resources (wells) that are at a certain z-height.

Because a Trough has no child (or parent), we have to define the `through_base_to_container_base` as a separate attribute. The alternative would be a "trough-parent and trough-child"-like structure, where the trough-child (the actual `Container`) is raised wrt trough-parent.

As you mentioned though, for wells, we should be taking into account the well bottom material thickness. I see that as _roughly_ analogous to `through_base_to_container_base`.

Reply:
> As you mentioned though, for wells, we should be taking into account the well bottom material thickness. I see that as _roughly_ analogous to `through_base_to_container_base`.

That is incorrect here: I've quickly made some rough drawings to visualise what the trough actually looks like:

![Hamilton_trough_200ml](https://github.com/PyLabRobot/pylabrobot/assets/122165124/934a8aa5-0d40-4231-b8cd-9e2257d7b4ab)

I highlighted the `Trough.through_base_to_container_base` with the actual measurement I took for it, 1.2 mm.

Since the "TroughCarrierSite".pedestal_z_height = 1.5 mm, the trough will therefore sit with its container.base directly on top of the "TroughCarrierSite".pedestal, while the "Trough.skirt" is hanging in the air leaving a 0.3 mm gap below it.

This is exactly like a plate with just one well.

Reply:
This is actually not different from what I thought 😅 What exactly is wrong here? Do you want to update this PR? I can revert so you can update, then I can remerge.

I realize my original phasing was not clear. This is not analogous in the physical geometry of these resources. I was just saying these are properties of "single-class" Resources and cannot be modeled as relations between resources. The well-dz is modeled as the relation between resources (namely in Well.location).

Reply:
> Because a Trough has no child (or parent), we have to define the through_base_to_container_base as a separate attribute. The alternative would be a "trough-parent and trough-child"-like structure, where the trough-child (the actual Container) is raised wrt trough-parent.

Yes, this is becoming confusing. Because a container is currently essentially a "paperbox", i.e. an object with no material_thickness associated with itself, i.e. all walls are believed to be 0mm wide, if I understand this correctly.

This is fine for wells because they sit inside the bigger plate - with the notable exception of their z-height, which is part of the big problem with `Plate` that we are trying to fix atm.
But in the case of troughs this does not really works as you can see above, because not only do we need to account for the material_thickness (at least in z) but here we also have to account for the "trough.child" / "container inside the trough" to be located at an offset to the left-front-bottom / origin of the trough.

Reply:
> Do you want to update this PR? I can revert so you can update, then I can remerge.

> I realize my original phasing was not clear. This is not analogous in the physical geometry of these resources.

No, I think we can leave this definition for now because I updated `Trough` to include
- `material_z_thickness=1.5 &`
- `through_base_to_container_base = 1.2,`

-> With this information people can aspirate/dispense with `offsets = [Coordiante(x, y, through_base_to_container_base + material_z_thickness]` which will then correctly target the bottom of this trough.

If we find this is cumbersome we can figure out a neater way in the future.

Reply:
I think in lh.aspirate or even STAR.aspirate & co we can incorporate this. Specifying offsets for every operation is not neat, that is something we can easily do automatically.